spotlessApply and spotlessCheck in the same build fails in 4.x

I have this in my build.gradle.kts:
```kotlin
defaultTasks("spotlessApply", "build")
```
so I always run formatter before build. It worked up until version 4.0.0. Now plugin formats everything but fails at the end:
```
> Task :spotlessJavaCheck FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':spotlessJavaCheck'.
> The following files had format violations:
      src/test/java/se/inoviagroup/v2t/gateway/AbstractSpringBootTest.java
          
  Run './gradlew :spotlessApply' to fix these violations.
```

If I run build again - it passes, since everything was formatted previously.

It worked on gradle 6.4.1 and spotless gradle plugin 3.30.0.

--- a/plugin-gradle/CHANGES.md
+++ b/plugin-gradle/CHANGES.md
@@ -3,7 +3,10 @@
 We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (starting after version `3.27.0`).
 
 ## [Unreleased]
-* Support for ktfmt in KotlinGradleExtension ([#583](https://github.com/diffplug/spotless/pull/583))
+### Added
+* Support for ktfmt in KotlinGradleExtension. ([#583](https://github.com/diffplug/spotless/pull/583))
+### Fixed
+* Users can now run `spotlessCheck` and `spotlessApply` in the same build. ([#584](https://github.com/diffplug/spotless/pull/584))
 
 ## [4.0.1] - 2020-05-21
 ### Fixed

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/FormatExtension.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/FormatExtension.java
@@ -635,7 +635,7 @@ public class FormatExtension {
 		// create the apply task
 		SpotlessApply applyTask = root.project.getTasks().create(taskName, SpotlessApply.class);
 		applyTask.setSpotlessOutDirectory(spotlessTask.getOutputDirectory());
-		applyTask.source = spotlessTask;
+		applyTask.linkSource(spotlessTask);
 		applyTask.dependsOn(spotlessTask);
 
 		return applyTask;

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessApply.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessApply.java
@@ -30,7 +30,14 @@ import org.gradle.api.tasks.PathSensitivity;
 import org.gradle.api.tasks.TaskAction;
 
 public class SpotlessApply extends DefaultTask {
-	SpotlessTask source;
+	private SpotlessTask source;
+
+	/** Bidirectional link between Apply and Spotless allows check to know if Apply ran or not. */
+	void linkSource(SpotlessTask source) {
+		this.source = source;
+		source.applyTask = this;
+	}
+
 	private File spotlessOutDirectory;
 
 	@PathSensitive(PathSensitivity.RELATIVE)

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessCheck.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessCheck.java
@@ -54,6 +54,9 @@ public class SpotlessCheck extends DefaultTask {
 		ConfigurableFileTree files = getProject().fileTree(spotlessOutDirectory);
 		if (files.isEmpty()) {
 			getState().setDidWork(source.getDidWork());
+		} else if (getProject().getGradle().getTaskGraph().hasTask(source.applyTask)) {
+			// if our matching apply has already run, then we don't need to do anything
+			getState().setDidWork(false);
 		} else {
 			List<File> problemFiles = new ArrayList<>();
 			files.visit(new FileVisitor() {

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessCheck.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessCheck.java
@@ -49,12 +49,20 @@ public class SpotlessCheck extends DefaultTask {
 		this.spotlessOutDirectory = spotlessOutDirectory;
 	}
 
+	public void performActionTest() throws Exception {
+		performAction(true);
+	}
+
 	@TaskAction
 	public void performAction() throws Exception {
+		performAction(false);
+	}
+
+	private void performAction(boolean isTest) {
 		ConfigurableFileTree files = getProject().fileTree(spotlessOutDirectory);
 		if (files.isEmpty()) {
 			getState().setDidWork(source.getDidWork());
-		} else if (getProject().getGradle().getTaskGraph().hasTask(source.applyTask)) {
+		} else if (!isTest && getProject().getGradle().getTaskGraph().hasTask(source.applyTask)) {
 			// if our matching apply has already run, then we don't need to do anything
 			getState().setDidWork(false);
 		} else {

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessExtension.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessExtension.java
@@ -249,7 +249,6 @@ public class SpotlessExtension {
 	 * - "spotless{FormatName}Check" will depend on the main spotless task in `check` mode
 	 * - "spotless{FormatName}Apply" will depend on the main spotless task in `apply` mode
 	 */
-	@SuppressWarnings("rawtypes")
 	private void createFormatTasks(String name, FormatExtension formatExtension) {
 		// create the SpotlessTask
 		String taskName = EXTENSION + SpotlessPlugin.capitalize(name);
@@ -268,8 +267,11 @@ public class SpotlessExtension {
 
 		SpotlessApply applyTask = project.getTasks().create(taskName + APPLY, SpotlessApply.class);
 		applyTask.setSpotlessOutDirectory(spotlessTask.getOutputDirectory());
-		applyTask.source = spotlessTask;
+		applyTask.linkSource(spotlessTask);
 		applyTask.dependsOn(spotlessTask);
+
+		// if the user runs both, make sure that apply happens first,
+		checkTask.mustRunAfter(applyTask);
 
 		// set the filePatterns property
 		project.afterEvaluate(unused -> {

--- a/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java
+++ b/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java
@@ -53,6 +53,15 @@ import com.diffplug.spotless.PaddedCell;
 
 @CacheableTask
 public class SpotlessTask extends DefaultTask {
+	SpotlessApply applyTask;
+
+	/** @deprecated internal use only, allows coordination between check and apply when they are in the same build */
+	@Internal
+	@Deprecated
+	public SpotlessApply getApplyTask() {
+		return applyTask;
+	}
+
 	// set by SpotlessExtension, but possibly overridden by FormatExtension
 	protected String encoding = "UTF-8";
 

--- a/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/DiffMessageFormatterTest.java
+++ b/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/DiffMessageFormatterTest.java
@@ -39,21 +39,15 @@ import com.diffplug.spotless.extra.integration.DiffMessageFormatter;
 public class DiffMessageFormatterTest extends ResourceHarness {
 
 	private class Bundle {
-		String name;
 		Project project = TestProvisioner.gradleProject(rootFolder());
 		File file;
-		File outputFile;
 		SpotlessTask task;
 		SpotlessCheck check;
-		SpotlessApply apply;
 
 		Bundle(String name) throws IOException {
-			this.name = name;
 			file = setFile("src/test." + name).toContent("CCC");
 			task = createFormatTask(name);
 			check = createCheckTask(name, task);
-			apply = createApplyTask(name, task);
-			outputFile = new File(task.getOutputDirectory() + "/src", file.getName());
 		}
 
 		private SpotlessTask createFormatTask(String name) {
@@ -70,13 +64,6 @@ public class DiffMessageFormatterTest extends ResourceHarness {
 			return task;
 		}
 
-		private SpotlessApply createApplyTask(String name, SpotlessTask source) {
-			SpotlessApply task = project.getTasks().create("spotless" + SpotlessPlugin.capitalize(name) + "Apply", SpotlessApply.class);
-			task.source = source;
-			task.setSpotlessOutDirectory(source.getOutputDirectory());
-			return task;
-		}
-
 		String checkFailureMsg() {
 			try {
 				check();
@@ -84,21 +71,6 @@ public class DiffMessageFormatterTest extends ResourceHarness {
 			} catch (Exception e) {
 				return e.getMessage();
 			}
-		}
-
-		void diagnose() throws IOException {
-			SpotlessDiagnoseTask diagnose = project.getTasks().create("spotless" + SpotlessPlugin.capitalize(name) + "Diagnose", SpotlessDiagnoseTask.class);
-			diagnose.source = task;
-			diagnose.performAction();
-		}
-
-		void format() throws Exception {
-			execute(task);
-		}
-
-		void apply() throws Exception {
-			execute(task);
-			apply.performAction();
 		}
 
 		void check() throws Exception {

--- a/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/DiffMessageFormatterTest.java
+++ b/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/DiffMessageFormatterTest.java
@@ -48,6 +48,7 @@ public class DiffMessageFormatterTest extends ResourceHarness {
 			file = setFile("src/test." + name).toContent("CCC");
 			task = createFormatTask(name);
 			check = createCheckTask(name, task);
+			createApplyTask(name, task);
 		}
 
 		private SpotlessTask createFormatTask(String name) {
@@ -64,6 +65,13 @@ public class DiffMessageFormatterTest extends ResourceHarness {
 			return task;
 		}
 
+		private SpotlessApply createApplyTask(String name, SpotlessTask source) {
+			SpotlessApply task = project.getTasks().create("spotless" + SpotlessPlugin.capitalize(name) + "Apply", SpotlessApply.class);
+			task.linkSource(this.task);
+			task.setSpotlessOutDirectory(source.getOutputDirectory());
+			return task;
+		}
+
 		String checkFailureMsg() {
 			try {
 				check();
@@ -75,7 +83,7 @@ public class DiffMessageFormatterTest extends ResourceHarness {
 
 		void check() throws Exception {
 			execute(task);
-			check.performAction();
+			check.performActionTest();
 		}
 	}
 

--- a/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/PaddedCellTaskTest.java
+++ b/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/PaddedCellTaskTest.java
@@ -107,7 +107,7 @@ public class PaddedCellTaskTest extends ResourceHarness {
 
 		void check() throws Exception {
 			execute(task);
-			check.performAction();
+			check.performActionTest();
 		}
 	}
 

--- a/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/PaddedCellTaskTest.java
+++ b/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/PaddedCellTaskTest.java
@@ -76,7 +76,7 @@ public class PaddedCellTaskTest extends ResourceHarness {
 
 		private SpotlessApply createApplyTask(String name, SpotlessTask source) {
 			SpotlessApply task = project.getTasks().create("spotless" + SpotlessPlugin.capitalize(name) + "Apply", SpotlessApply.class);
-			task.source = source;
+			task.linkSource(source);
 			task.setSpotlessOutDirectory(source.getOutputDirectory());
 			return task;
 		}

--- a/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/SpecificFilesTest.java
+++ b/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/SpecificFilesTest.java
@@ -16,23 +16,38 @@
 package com.diffplug.gradle.spotless;
 
 import java.io.IOException;
+import java.util.Locale;
 
-import org.gradle.testkit.runner.*;
+import org.gradle.testkit.runner.GradleRunner;
+import org.gradle.testkit.runner.UnexpectedBuildFailure;
 import org.junit.Ignore;
 import org.junit.Test;
 
+import com.diffplug.common.base.StandardSystemProperty;
+
 public class SpecificFilesTest extends GradleIntegrationTest {
+	private static boolean isWindows() {
+		return StandardSystemProperty.OS_NAME.value().toLowerCase(Locale.US).contains("win");
+	}
+
+	private static String regexWinSafe(String input) {
+		return isWindows() ? input.replace("/", "\\\\") : input;
+	}
+
 	private String testFilePath(int number) {
 		return testFilePath(number, true);
 	}
 
 	private String testFilePath(int number, boolean absolute) {
 		String relPath = "src/main/java/test" + number + ".java";
+		String returnValue;
 		if (absolute) {
-			return rootFolder() + "/" + relPath;
+			returnValue = rootFolder().getAbsolutePath().replace('\\', '/') + "/" + relPath;
 		} else {
-			return relPath;
+			returnValue = relPath;
 		}
+		// regex-escape on windows;
+		return regexWinSafe(returnValue);
 	}
 
 	private String fixture() {
@@ -141,7 +156,7 @@ public class SpecificFilesTest extends GradleIntegrationTest {
 	@Test
 	public void regexp() throws IOException {
 		createBuildScript();
-		integration(".*/src/main/java/test(1|3).java", true, false, true);
+		integration(regexWinSafe(".*/src/main/java/test(1|3).java"), true, false, true);
 	}
 
 	@Test(expected = UnexpectedBuildFailure.class)

--- a/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/UpToDateTest.java
+++ b/plugin-gradle/src/test/java/com/diffplug/gradle/spotless/UpToDateTest.java
@@ -98,4 +98,11 @@ public class UpToDateTest extends GradleIntegrationTest {
 		applyIsUpToDate(false);
 		applyIsUpToDate(true);
 	}
+
+	@Test
+	public void checkAndApply() throws IOException {
+		writeBuildFile();
+		setFile("README.md").toContent("ABC");
+		gradleRunner().withArguments("spotlessCheck", "spotlessApply").build();
+	}
 }

